### PR TITLE
Vaibhav/milestone [DO NOT MERGE]

### PR DESCRIPTION
**Pull request by VAIBHAVJINDAL3012**
# Description

Please provide a detailed description of what was done in this PR
This PR includes the changes which involves using the endBlock hash instead of rootHash in the milestone.

# Changes

- [ ] Bugfix (non-breaking change that solves an issue)
- [ ] Hotfix (change that solves an urgent issue, and requires immediate attention)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (change that is not backwards-compatible and/or changes current functionality)
- [ ] Changes only for a subset of nodes

# Breaking changes

Please complete this section if any breaking changes have been made, otherwise delete it

# Nodes audience

In case this PR includes changes that must be applied only to a subset of nodes, please specify how you handled it (e.g. by adding a flag with a default value...)

# Checklist

- [ ] I have added at least 2 reviewer or the whole pos-v1 team
- [ ] I have added sufficient documentation in code
- [ ] I will be resolving comments - if any - by pushing each fix in a separate commit and linking the commit hash in the comment reply

# Cross repository changes

- [ ] This PR requires changes to heimdall
    - In case link the PR here:
- [ ] This PR requires changes to matic-cli
    - In case link the PR here:

## Testing

- [ ] I have added unit tests
- [ ] I have added tests to CI
- [ ] I have tested this code manually on local environment
- [ ] I have tested this code manually on remote devnet using express-cli
- [ ] I have tested this code manually on mumbai
- [ ] I have created new e2e tests into express-cli

### Manual tests

Please complete this section with the steps you performed if you ran manual tests for this functionality, otherwise delete it

# Additional comments

Please post additional comments in this section if you have them, otherwise delete it

### CODE REVIEW

File: eth/bor_api_backend.go [hunk 1]
```diff
@@ -48,35 +49,31 @@ func (b *EthAPIBackend) GetVoteOnRootHash(ctx context.Context, starBlockNr uint6
...
 	if err != nil {
-		return false, err
+		return false, errEndBlock
 	}
 
-	if !b.eth.Miner().Mining() {
-		return localRootHash == rootHash, nil
+	localEndBlockHash := localEndBlock.Hash().String()
+
+	if err != nil {
```
Comment (JekaMas) [inline eth/bor_api_backend.go:59]: do we need this check?

File: eth/bor_api_backend.go [hunk 2]
```diff
@@ -48,35 +49,31 @@ func (b *EthAPIBackend) GetVoteOnRootHash(ctx context.Context, starBlockNr uint6
...
-		return false, errors.New("Previous sprint is still in locked state")
+		return false, errors.New("Whitelisted number or locked sprint number is more than the received end block number")
 	}
 
-	if localRootHash != rootHash[2:] {
+	if localEndBlockHash != endBlockHash {
 		downloader.UnlockMutex(false, "", common.Hash{})
-
-		return false, errors.New("RootHash mismatch+" + localRootHash + rootHash)
+		return false, errors.New("EndBlockHash mismatch")
```
Comment (JekaMas) [inline eth/bor_api_backend.go:73]: could we have this error as a variable/constant + add error context with fmt.Errorf("%w: hash1 %s, hash2 %s", ...)?

File: eth/bor_checkpoint_verifier.go [hunk 3]
```diff
@@ -32,32 +32,46 @@ var (
...
+		}
+	} else {
+		//in case of milestone(isCheckpoint==false) get the hash of endBlock
+		block, err := handler.ethAPI.GetBlockByNumber(ctx, rpc.BlockNumber(end), false)
+		if err != nil {
+			log.Debug("Failed to get end block hash while whitelisting milestone", "err", err)
+			return hash, errEndBlock
+		}
+
+		localHash = fmt.Sprintf("%v", block["hash"])
```
Comment (JekaMas) [inline eth/bor_checkpoint_verifier.go:70]: do we really need fmt.Sprintf for that?

Comment (JekaMas) [reply eth/bor_checkpoint_verifier.go:70]: block["hash"].Hash/String/Hex() depending on what you need.

**Review by JekaMas — APPROVED** [review]

File: eth/bor_api_backend.go (hunk 1 above)
Comment (VAIBHAVJINDAL3012) [reply eth/bor_api_backend.go:59]: No, Forgot to remove it while reframing the code

Comment (VAIBHAVJINDAL3012) [reply eth/bor_api_backend.go:59]: Removed it now.

File: eth/bor_checkpoint_verifier.go (hunk 3 above)
Comment (VAIBHAVJINDAL3012) [reply eth/bor_checkpoint_verifier.go:80]: Not working

File: ethclient/bor_ethclient.go [hunk 4]
```diff
@@ -17,6 +17,16 @@ func (ec *Client) GetRootHash(ctx context.Context, startBlockNumber uint64, endB
 	return rootHash, nil
 }
 
+// GetRootHash returns the merkle root of the block headers
+func (ec *Client) GetVoteOnHash(ctx context.Context, startBlockNumber uint64, endBlockNumber uint64, hash string, milestoneID string) (bool, error) {
+	var value bool
+	if err := ec.c.CallContext(ctx, &value, "eth_getVoteOnHash", startBlockNumber, endBlockNumber, hash, milestoneID); err != nil {
```
Comment (manav2401) [inline ethclient/bor_ethclient.go:23]: It should be `bor_getVoteOnHash`, right?

File: eth/bor_checkpoint_verifier.go [hunk 5]
```diff
@@ -1,61 +1,153 @@
...
+		rewind(eth, head, rewindTo)
+
+		eth.Miner().Start(eth.etherbase)
+	} else {
+		rewind(eth, head, rewindTo)
+	}
+}
+
+func rewind(eth *Ethereum, head uint64, rewindTo uint64) {
+	log.Warn("Rewinding chain to :", rewindTo, "block number")
```
Comment (manav2401) [inline eth/bor_checkpoint_verifier.go:144]: This will throw an error because the order is interchanged. Let's make the log something like
```
log.Warn("Rewinding chain due to milestone whitelist mismatch", "to", rewindTo)
```
Also, please try to add some context to logs (like the reason of rewind) so that it becomes convenient and understandable for us as well as node operators.

File: eth/bor_checkpoint_verifier.go [hunk 6]
```diff
@@ -1,61 +1,153 @@
...
+	err := eth.blockchain.SetHead(rewindTo)
 
-		return hash, nil
+	if err != nil {
+		log.Error("Error while rewinding the chain to", "Block Number", rewindTo, "Error", err)
```
Comment (manav2401) [inline eth/bor_checkpoint_verifier.go:148]: Update the log here as well. The tag for error is `"err", err`. Let's use that everywhere because while searching on indexed data on datadog, it will be useful to have uniform logs. Update other places as well (if any)
```
log.Error("Error while rewinding the chain", "to", rewindTo, "err", err)
```

File: eth/downloader/whitelist/service.go [hunk 7]
```diff
@@ -3,133 +3,152 @@ package whitelist
...
-		return true, nil
+	order, list, err := rawdb.ReadFutureMilestoneList(db)
+	if err != nil {
+		order = make([]uint64, 0)
+		list = make(map[uint64]common.Hash)
 	}
 
-	return false, ErrCheckpointMismatch
+	return &Service{
+		//fixme: try to restore from DB
```
Comment (manav2401) [inline eth/downloader/whitelist/service.go:56]: Is this for future optimisations?

File: eth/downloader/whitelist/service.go [hunk 8]
```diff
@@ -3,133 +3,152 @@ package whitelist
...
+			finality[*rawdb.Checkpoint]{
+				doExist:  checkpointDoExist,
+				Number:   checkpointNumber,
+				Hash:     checkpointHash,
+				interval: 256,
+				db:       db,
+			},
+		},
+
+		//fixme: try to restore from DB
```
Comment (manav2401) [inline eth/downloader/whitelist/service.go:67]: Same comment as above.

File: eth/ethconfig/config.go [hunk 9]
```diff
@@ -33,6 +33,8 @@ import (
 	"github.com/ethereum/go-ethereum/consensus/bor/heimdall" //nolint:typecheck
 	"github.com/ethereum/go-ethereum/consensus/bor/heimdall/span"
 	"github.com/ethereum/go-ethereum/consensus/bor/heimdallgrpc"
+
+	//"github.com/ethereum/go-ethereum/consensus/bor/heimdallgrpc"
```
Comment (manav2401) [inline eth/ethconfig/config.go:37]: Let's remove this line.

File: params/config.go [hunk 10]
```diff
@@ -229,7 +229,8 @@ var (
 		MuirGlacierBlock:    nil,
 		BerlinBlock:         big.NewInt(4_460_644),
 		LondonBlock:         big.NewInt(5_062_605),
-		ArrowGlacierBlock:   nil,
+
+		ArrowGlacierBlock: nil,
```
Comment (manav2401) [inline params/config.go:233]: Let's undo this change.

File: eth/bor_checkpoint_verifier.go (hunk 5 above)
Comment (VAIBHAVJINDAL3012) [reply eth/bor_checkpoint_verifier.go:144]: Okay!

File: eth/downloader/whitelist/service.go (hunk 7 above)
Comment (VAIBHAVJINDAL3012) [reply eth/downloader/whitelist/service.go:56]: Forgot to remove the comments....Actually restoring it from Db

File: core/blockchain_test.go [hunk 11]
```diff
@@ -1891,9 +1891,13 @@ func TestLowDiffLongChain(t *testing.T) {
 	db := rawdb.NewMemoryDatabase()
 	genesis := (&Genesis{BaseFee: big.NewInt(params.InitialBaseFee)}).MustCommit(db)
 
+	//Using TempTriesInMemory variable instead of DefaultTempInTries because changing the
+	//value of DefaultTempInTries to 1024 is failing the test.
+	TempTriesInMemory := 128
```
[the remark this answers is not in the document]
Comment (VAIBHAVJINDAL3012) [reply core/blockchain_test.go:1896]: Not a temporary fix, We need to change the value of TriesInMemory to 1024, But these test fail with 1024 value because in these tests they are trying  to import the DefaultTriesInMemory(value) chain.
But please take one look into it.

File: core/blockchain.go [hunk 12]
```diff
@@ -145,7 +145,7 @@ var DefaultCacheConfig = &CacheConfig{
 	TrieTimeLimit:  5 * time.Minute,
 	SnapshotLimit:  256,
 	SnapshotWait:   true,
-	TriesInMemory:  128,
+	TriesInMemory:  1024,
```
[the remark this answers is not in the document]
Comment (VAIBHAVJINDAL3012) [reply core/blockchain.go:148]: No, We need to take it with milestone change only.....We can discuss it with a team.

File: eth/bor_checkpoint_verifier.go (hunk 6 above)
Comment (VAIBHAVJINDAL3012) [reply eth/bor_checkpoint_verifier.go:148]: What do mean by tag here?

Comment (manav2401) [reply eth/bor_checkpoint_verifier.go:148]: Sorry for the wrong wordings. I meant the way we pass args for error. It's generally `"err", err` everywhere.

File: core/blockchain.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (manav2401) [reply core/blockchain.go:148]: Okay

File: core/blockchain_test.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (manav2401) [reply core/blockchain_test.go:1896]: Okay
